### PR TITLE
add inter-component communication vision document

### PR DESCRIPTION
**Pull request by lukaskollmer**
This PR adds a vision document for inter-component communication.

You can find the rendered markdown file [here](https://github.com/Apodini/Apodini/blob/documentation/inter-component-communication/Documentation/Components/Inter-Component%20Communication.md).

This is intended to be used with web services deployed in a distributed fashion, which will be the subject of another vision document.

### CODE REVIEW

File: Documentation/Components/Inter-Component Communication.md [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+## The `InvocableHandler` protocol
+
+```swift
+public protocol InvocableHandler: IdentifiableHandler where Response: Decodable {
+    typealias ParametersStorageProtocol = InvocableHandlerParametersStorageProtocol
+    associatedtype ParametersStorage: ParametersStorageProtocol = _EmptyParametersStorage<Self> where ParametersStorage.HandlerType == Self
+}
+```
+
+An `InvocableHandler` is an `IdentifiableHandler` whose `Response` type conforms to `Decodable`, and which optionally defines a custom type for passing parameters to a remote invcation (see below for an explanation of how parameter passsing works).
```
Comment (awoc) [inline Documentation/Components/Inter-Component Communication.md]:
```suggestion
An `InvocableHandler` is an `IdentifiableHandler` whose `Response` type conforms to `Decodable`, and which optionally defines a custom type for passing parameters to a remote invocation (see below for an explanation of how parameter passsing works).
```

File: Documentation/Components/Inter-Component Communication.md [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+    }
+}
+```
+
+
+
+## The Remote Invocation Manager
+
+
+The invocation manager acts as a mediator overseeing all inter-handler interactions for a handler. Each handler which wants to other handlers defines its own instance of the invocation manager.
```
Comment (awoc) [inline Documentation/Components/Inter-Component Communication.md]:
```suggestion
The invocation manager acts as a mediator overseeing all inter-handler interactions for a handler. Each handler which wants to interact with other handlers defines its own instance of the invocation manager.
```

File: Documentation/Components/Inter-Component Communication.md [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+
+
+### What's an interaction?
+
+An interaction is an event where date are transferred between two handlers, and which was initiated within one of the two handler's `handle()` functions.
```
Comment (awoc) [inline Documentation/Components/Inter-Component Communication.md:80]:
```suggestion
An interaction is an event where data is transferred between two handlers, and which was initiated within one of the two handler's `handle()` functions.
```

I would use "data is" but "data are" is apparently also correct: https://www.theguardian.com/news/datablog/2010/jul/16/data-plural-singular

File: Documentation/Components/Inter-Component Communication.md [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+  - These are interactions where sender and target are running in different processed (potentially even on different machines and different networks)
+  - Remote interactions are available when running a web service as a distributed system (ie, a system of distinct nodes, each of which implement a subset of the web service's exported endpoints)
+  - In this case, the interaction is realised by finding the correct node within our distributed system, invoking the target component there, and returning the result back to the sender
+
+
+### What does the invocation manager do?
+
+The invocation manager is responsible for deciding whether the target component should be invoked locally or remotely, thus acting as a dispatcher. It is also responsible for realising the interaction.
+
+The invocation manager has access to the full structure of the web service (eg via the `WebServiceModel` object) and the deployment structure structure (ie information about the number of processes running the web service, the endpoints implemented by each process, etc).
```
Comment (awoc) [inline Documentation/Components/Inter-Component Communication.md]:
```suggestion
The invocation manager has access to the full structure of the web service (eg via the `WebServiceModel` object) and the deployment structure (ie information about the number of processes running the web service, the endpoints implemented by each process, etc).
```

File: Documentation/Components/Inter-Component Communication.md [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+(`_EmptyParametersStorage ` is a non-initialisable type which a Handler can use to opt in to the array-based parameter passing. For `InvocableHandler`s which do not specify a `ParametersStorage` type, the parameter storage defaults to this type.)
+
+Similar to how the `ParametersStorage` consisted of the parameter values alongside an array mapping the key path to a value (within the parameters storage) to the key path of its respective `@Parameter` (within the Handler), the `CollectedParameter`s expected by this function consist of the parameter value and the key path to the handler's `@Parameter`.
+
+Entries for `@Parameter`s which define a default value may be omitted from the parameters array.  
+If the parameters array contains multiple entries for the same `@Parameter` key path, the last one is used.  
+Incomplete parameter lists (eg missing parameters) result in a run-time error.
+
+The advantage of this parameter-passing approach is that the handler doesn't have to write boilerplate code.  
+The downside however is the loss of compile-time well-formedness checking.
```
Comment (awoc) [inline Documentation/Components/Inter-Component Communication.md]:
```suggestion
The downside, however, is the loss of compile-time well-formedness checking.
```

File: Documentation/Components/Inter-Component Communication.md [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+
+# Inter-Component Communication
+
+
+
+
+## Summary
+
+- The APIs described in this document are used to access, from within a handler's `handle()` function, the functionality provided by another handler
+- The `InvocableHandler` protocol, which interits from `IdentifiableHandler`, is used to denote `Handler`s which can be remotely invoked (ie, from within a `Handler.handle()` function)
```
Comment (awoc) [inline Documentation/Components/Inter-Component Communication.md]:
```suggestion
- The `InvocableHandler` protocol, which inherits from `IdentifiableHandler`, is used to denote `Handler`s which can be remotely invoked (ie, from within a `Handler.handle()` function)
```

**Comment by lukaskollmer** [comment]
@awoc thanks for the feedback. I incorporated all of your suggestions, except the one regarding "data is" vs "data are". (This is something I did in fact look into. From a language point of view, both are allowed, and I prefer the latter one.)

**Review by PSchmiedmayer — COMMENTED** [review]
@lukaskollmer Thanks for the proposal, I enjoyed reading it and seeing how the ideas from our thesis proposal have evolved in this proposal. I like the elegance of using a dedicated `ParametersStorage` struct to offer a `InvocableHandler` defined parameter passing and the flexibility of using an array-based parameter passing. I have noted down a few comments and thoughts and would suggest that we discuss and evaluate them together tomorrows thesis meeting in detail and share the results here too.

**Review by PSchmiedmayer — APPROVED** [review]
Approved after the discussion in the thesis meeting.

**Pull request merged**